File: app/src/main/java/com/example/mealrecipeapp/ui/signin/SignInFragment.java
package com.example.mealrecipeapp.ui.signin;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.credentials.CredentialManager;
import androidx.credentials.CredentialManagerCallback;
import androidx.credentials.GetCredentialRequest;
import androidx.credentials.GetCredentialResponse;
import androidx.credentials.exceptions.GetCredentialException;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;

import com.example.mealrecipeapp.MealRecipeApp;
import com.example.mealrecipeapp.databinding.FragmentSignInBinding;
import com.example.mealrecipeapp.di.AppContainer;
import com.example.mealrecipeapp.ui.dialog.LoadinDialog;
import com.google.android.libraries.identity.googleid.GetSignInWithGoogleOption;
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential;

public class SignInFragment extends Fragment {

    private FragmentSignInBinding binding;

    private SignInViewModel signInViewModel;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentSignInBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        AppContainer appContainer = ((MealRecipeApp) requireActivity().getApplication()).appContainer;
        signInViewModel = new ViewModelProvider(this, appContainer.viewModelFactory).get(SignInViewModel.class);

        CredentialManager credentialManager = CredentialManager.create(requireContext());
        GetSignInWithGoogleOption googleIdOption = new GetSignInWithGoogleOption
                .Builder("347653735490-n45emv6v7jdvat65jc48gruihov3ct37.apps.googleusercontent.com")
                .build();
        GetCredentialRequest request = new GetCredentialRequest.Builder()
                .addCredentialOption(googleIdOption)
                .build();

        binding.signInButton.setIconTint(null);

        binding.signInButton.setOnClickListener(onClick -> credentialManager.getCredentialAsync(
            requireActivity(),
            request,
            null,
            requireActivity().getMainExecutor(),
            new CredentialManagerCallback<GetCredentialResponse, GetCredentialException>() {
                @Override
                public void onResult(GetCredentialResponse getCredentialResponse) {
                    Bundle credentialData = getCredentialResponse.getCredential().getData();

                    GoogleIdTokenCredential googleIdTokenCredential = GoogleIdTokenCredential.createFrom(credentialData);

                    String email = googleIdTokenCredential.getId();
                    String name = googleIdTokenCredential.getDisplayName();
                    String image = "";
                    if (googleIdTokenCredential.getProfilePictureUri() != null) {
                        image = googleIdTokenCredential.getProfilePictureUri().toString();
                    }

                    signInViewModel.saveUser(email, name, image);
                }

                @Override
                public void onError(@NonNull GetCredentialException e) {

                }
            }
        ));

        observeSignInResult();
    }

    private void observeSignInResult() {
        LoadinDialog loadinDialog = new LoadinDialog(requireContext());
        signInViewModel.getSignInResult().observe(getViewLifecycleOwner(), resource -> {
            switch (resource.getStatus()) {
                case SUCCESS:
                    loadinDialog.dismiss();
                    NavHostFragment.findNavController(this).navigate(SignInFragmentDirections.actionSignInFragmentToHomeFragment());
                    break;
                case ERROR:
                    loadinDialog.dismiss();
                    Toast.makeText(getActivity(), resource.getMessage(), Toast.LENGTH_SHORT).show();
                    break;
                case LOADING:
                    loadinDialog.show();
                    break;
            }
        });
    }
}

File: app/src/main/java/com/example/mealrecipeapp/ui/signin/SignInFragment.kt
package com.example.mealrecipeapp.ui.signin

import android.app.Activity.RESULT_OK
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.findNavController
import com.example.mealrecipeapp.MealRecipeApp
import com.example.mealrecipeapp.databinding.FragmentSignInBinding
import com.example.mealrecipeapp.ui.dialog.LoadingDialog
import com.example.mealrecipeapp.utils.Constants
import com.example.mealrecipeapp.utils.Resource
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.OAuthProvider


class SignInFragment : Fragment() {
    private var _binding: FragmentSignInBinding? = null
    private val binding get() = _binding!!

    private lateinit var signInViewModel: SignInViewModel

    private val auth = FirebaseAuth.getInstance()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentSignInBinding.inflate(inflater, container, false)


        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val appContainer = (requireActivity().application as MealRecipeApp).appContainer
        signInViewModel = ViewModelProvider(this, appContainer.viewModelFactory)[SignInViewModel::class.java]

        val gso = GoogleSignInOptions.Builder()
            .requestIdToken(Constants.GOOGLE_CLIENT_ID)
            .build()

        val googleSignInClient = GoogleSignIn.getClient(requireContext(), gso)

        val signInGoogleIntent = googleSignInClient.signInIntent

        val launcher = registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
        ) { result: ActivityResult ->
            if (result.resultCode == RESULT_OK) {
                val task = GoogleSignIn.getSignedInAccountFromIntent(result.data)
                try {
                    val account = task.getResult(ApiException::class.java)
                    val credential = GoogleAuthProvider.getCredential(account.idToken, null)
                    auth.signInWithCredential(credential)
                        .addOnCompleteListener(requireActivity()) { credentialTask ->
                            if (credentialTask.isSuccessful) {
                                signInViewModel.saveUser()
                            } else {
                                Toast.makeText(requireContext(), "Error: Authentication Failed", Toast.LENGTH_SHORT).show()
                            }
                        }
                } catch (e: ApiException) {
                    Toast.makeText(requireContext(), "Error : ${e.message}", Toast.LENGTH_LONG).show()
                }
            }

            googleSignInClient.signOut()
        }

        binding.signInButton.iconTint = null
        binding.signInButton.setOnClickListener {
            launcher.launch(signInGoogleIntent)
        }

        binding.signInGithubButton.iconTint = null
        binding.signInGithubButton.setOnClickListener {
            signInWithGithub()
        }

        observeSignInResult()
    }

    private fun signInWithGithub() {
        val provider = OAuthProvider.newBuilder("github.com")
        provider.scopes = listOf("user:email")

        val pendingResultTask = auth.pendingAuthResult
        if (pendingResultTask != null) {
            pendingResultTask
                .addOnSuccessListener {
                    signInViewModel.saveUser()
                }
                .addOnFailureListener {
                    Toast.makeText(requireContext(), "Error : $it", Toast.LENGTH_LONG).show()
                }
        } else {
            auth.startActivityForSignInWithProvider(requireActivity(), provider.build())
                .addOnSuccessListener {
                    signInViewModel.saveUser()
                }
                .addOnFailureListener {
                    Toast.makeText(requireContext(), "Error : $it", Toast.LENGTH_LONG).show()
                }
        }
    }

    private fun observeSignInResult() {
        val loadingDialog = LoadingDialog(requireContext())
        signInViewModel.signInResult.observe(viewLifecycleOwner) { resource->
            when (resource) {
                is Resource.Success -> {
                    loadingDialog.dismiss()
                    findNavController().navigate(SignInFragmentDirections.actionSignInFragmentToHomeFragment())
                }

                is Resource.Error -> {
                    loadingDialog.dismiss()
                    Toast.makeText(activity, resource.message, Toast.LENGTH_SHORT).show()
                }

                is Resource.Loading -> loadingDialog.show()
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}